File: roman/roman.clj
(ns roman)

(def numerals [["M" 1000]
               ["CM" 900]
               ["D" 500]
               ["CD" 400]
               ["C" 100]
               ["XC" 90]
               ["L" 50]
               ["XL" 40]
               ["X" 10]
               ["IX" 9]
               ["V" 5]
               ["IV" 4]
               ["I" 1]])

(defn convert-recursive
  ([number]
   (convert-recursive number 0 ""))
  ([remainder numeral-index roman]
   (if (zero? remainder)
     roman
     (let [[symbol value] (nth numerals numeral-index)]
       (if (>= remainder value)
         (convert-recursive (- remainder value) numeral-index (str roman symbol))
         (convert-recursive remainder (inc numeral-index) roman))))))

(defn convert [number]
  "Simulates tail recursion by optimizing it into loop"
  (loop [remainder number, numeral-index 0, roman ""]
    (if (zero? remainder)
      roman
      (let [[symbol value] (nth numerals numeral-index)]
        (if (>= remainder value)
          (recur (- remainder value) numeral-index (str roman symbol))
          (recur remainder (inc numeral-index) roman))))))

;; (roman/convert-recursive 900)
;; (roman/convert 909)

File: roman/Roman.hs
type RomanAndArabic = (String, Int)

numerals :: [RomanAndArabic]
numerals =
  [ ("M", 1000),
    ("CM", 900),
    ("D", 500),
    ("CD", 400),
    ("C", 100),
    ("XC", 90),
    ("L", 50),
    ("XL", 40),
    ("X", 10),
    ("IX", 9),
    ("V", 5),
    ("IV", 4),
    ("I", 1)
  ]

roman :: Int -> [Char]
roman n
  | n <= 0 = "-"
  | otherwise = convert n 0
  where
    convert 0 _ = ""
    convert n i =
      let (numeral, value) = numerals !! i
       in if n >= value
            then numeral ++ convert (n - value) i
            else convert n (i + 1)

